### ISSUE TITLE
New language pairs

### ISSUE THREAD

**Issue by Path-Seeker** [quickmt/quickmt#21]
Hello!

First of all, thank you for all your hard work! I really enjoy using your models instead of NLLB; they are much faster and provide good quality.

I wonder how hard it would be to add several additional language pairs. I'm not a Python and ML/AI developer, but I managed to repeat the training from the old docs/training.md (https://github.com/quickmt/quickmt/blob/f6ad659ab1d8593d178ae9cbee2d30a726ed3b52/docs/training.md), with tons of bugs and issues on every step, but at least it works. I possess a pretty powerful PC, and it seems that I can handle the training in a reasonable amount of time (up to ~36 hours per model on my RTX 4080), unless some issues appear in the later stages of the training.

However I cannot evaluate whether it is possible to achieve decent quality for the language pairs I need with the available datasets:
Portuguese <-> Ukrainian
Portuguese <-> Russian
English <-> Ukrainian
German <-> Ukrainian
German <-> Russian
Russian <-> Ukrainian

Like I said, I can handle the training process, and I'm ok to publish datasets and models as a part of quickmt project, but I need help with scripts to collect datasets and properly run the training.

Would appreciate some feedback/thoughts on that.

Thanks again!

**Comment by radinplaid** [comment]
Cool! Thanks for your interest!

I'm currently working on improving the model training process. I really like eole and it worked well enough to create all of the current quickmt models but I wanted to experiment with creating my own training codebase. It's very much a work in progress, I plan to move it to the "quickmt" project space when it's a bit more mature, but you can check it out here: https://github.com/radinplaid/nmt_from_scratch

To evaluate whether the models achieve decent quality the [flores](https://huggingface.co/datasets/openlanguagedata/flores_plus) dataset is probably adequate. [WMT24pp](https://huggingface.co/datasets/google/wmt24pp) is probably better but is available for fewer languages. Previously I used this notebook to compare to NLLB and OPUS MT: https://github.com/quickmt/quickmt/blob/f6ad659ab1d8593d178ae9cbee2d30a726ed3b52/docs/notebooks/Benchmark.ipynb. For the current models I found a recipe that worked well enough and just ran with it. I'm sure there are tweaks that could improve quality/performance but hyper-parameter tuning is expensive :)

The code for downloading datasets uses [mtdata](https://github.com/thammegowda/mtdata), like you saw in the [training notebook](https://github.com/quickmt/quickmt/blob/f6ad659ab1d8593d178ae9cbee2d30a726ed3b52/docs/training.md). Are you having any specific problems? I'm happy to help.

Do you have any models you've already trained on huggingface? I'm happy to give them a try and provide feedback / help with the evaluation.

**Comment by Path-Seeker** [comment]
Thanks, I'll take a look at your repository. Do you have any estimations when you think it'll be mature enough to start real experiments with training new QuickMT models? And what's your goal here, except studying and experimenting?

Probably I explained incorrectly what I mean by "I cannot evaluate whether it is possible to achieve decent quality" - I meant that if there's a way to check if the training dataset is good enough. E.g. for the pair "German <-> Ukrainian" I may have such a small dataset that the training won't make sense at all. I see from your docs that you set specific datasets for dev, test, and train. Could you share why you chose these particular datasets? Are they the same for all QuickMT models, or do you adjust settings from model to model (for example, finding some datasets better than others)? The same question goes for training config (eole-config.yaml) - is it the same for all models, or does it vary? Probably that's what you meant by hyper-parameters.

Mostly, my struggle in repeating your process is because of the various versions of different libs conflict with each other. The versions I have for mtdata don't work with quickmt, and eole has its own dependencies, which conflict with all others (and I wasn't even able to install "flash-attn" properly), and so on and on. The "lid.176.bin" wasn't even there and I'm not sure that I have the correct version, and the entire "Clean Data a Bit" section of the docs looks like magic to me. Like I said, I'm not a Python developer, so maybe I'm just doing something wrong, but anyway it's not really related to QuickMT. I just wanted to repeat the process, try to train new models that I need, and maybe even publish them as QuickMT models.

**Comment by radinplaid** [comment]
I will try to solve all of the above issues, eventually, through a new 'quickmt-train' repository and package that will aim to make creating `quickmt` models easy and (hopefully) reproducible. 

The environment creation problem will be solved because I'll use Pytorch directly and the dependencies will be greatly simplified. Everything will work will work from a single config file (downloading data, cleaning, uploading data to huggingface, model training, model exporting, uploading the model to huggingface, ...). I'll try to get this out ASAP.

But to answer your other direct questions:

* It's just experimenting and studying - I have no commercial interest.
* Not all the configs are the same for existing quickmt models - I did some experimenting and made improvements along the way
* I tried to always use NTREX and flores-dev for the dev sets, but they were not available for all language pairs, so I chose sets that were the most test set like for these languages, and always had at least 2 datasets in the 'dev' split. Flores-devtest and wmt25pp I wanted to use for test sets, so tried to not have these in the train or dev split.
* `flash-attn` is not required - I never install it. It would give a boost in training speed in theory, but I never have the patience
* This is the link to lid.176.bin: https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin

Thanks for your interest and stay tuned - I'll post updates here as my training repo gets more mature

**Comment by radinplaid** [comment]
> Probably I explained incorrectly what I mean by "I cannot evaluate whether it is possible to achieve decent quality" - I meant that if there's a way to check if the training dataset is good enough. E.g. for the pair "German <-> Ukrainian" I may have such a small dataset that the training won't make sense at all.

I expect at least 10 - 20 million data samples will be required to get a good model. It might be possible to get a good model with fewer than 10 million samples but I imagine it would require tricks like [backtranslation](https://aclanthology.org/2023.findings-acl.518.pdf) or forward translation (possibly LLM distillation).

**Comment by Path-Seeker** [comment]
Thank you for the answers, it's very helpful! I'll keep you posted in case I manage to train some of the language pairs I mentioned I needed, and please don't hesitate to contact me if you need some additional computing power for training and testing some hypotheses.

**Comment by radinplaid** [comment]
I've made some progress! 

* UK->EN model trained: https://huggingface.co/datasets/quickmt/quickmt-train.uk-en
* EN->UK model almost done
* New training codebase progress: https://github.com/quickmt/quickmt-train
* Tutorial with some info on creating your own training corpus: https://quickmt.github.io/quickmt-train/corpus_creation/

If you end up trying to train a non-English centric model I'd be curious to hear how well it works out. My gut tells me it will be close to impossible to come close in quality to the many-to-many models (MADLAD, NLLB) and LLMs (Gemma 4, Aya Expanse, Tiny Aya, ...), but it should be easy to train a faster model.

**Comment by Path-Seeker** [comment]
Sounds impressive, great job! I'll give it a try in the next few days. And you'd be surprised how bad NLLB is with non-English pairs, and maybe I won't be able to overcome it, but being at least close to it with the quickmt performance and resource consumption would be great. I've also been thinking about enriching corpora for the non-English pairs by utilizing English pairs. E.g., if we have en<->pt corpora, and we have en<->uk corpora, I was thinking of trying to find the pairs with the same english variant and merge them into pt<->uk, to see how many such pairs there are and whether they increase the quality or not. But maybe it's not a new idea, and current non-english corpora already contain such pairs.
```
en <-> pt: "Hello" <-> "Olá";
en <-> uk: "Hello" <-> "Привіт";
```
Then obviously:
```
pt <-> uk: "Olá" <-> "Привіт"
```

**Comment by radinplaid** [comment]
Yes, that's a good idea! I think the "multi" corpora in opus try to do something like this e.g. [MultiParaCrawl](https://opus.nlpl.eu/datasets/MultiParaCrawl), [MultiCCAligned v1.1](https://opus.nlpl.eu/datasets/MultiCCAligned?pair=de&uk)

**Comment by radinplaid** [comment]
EN->UK model ready !

Looks pretty solid: https://huggingface.co/quickmt/quickmt-en-uk

It's tough to beat the "big" Opus models like https://huggingface.co/Helsinki-NLP/opus-mt-tc-big-en-zle. Especially the multi-lingual ones like that one. There definitely seem to be advantages to having multiple related languages mixed in and having backtranslated data like that one does.

**Comment by Path-Seeker** [comment]
> Yes, that's a good idea! I think the "multi" corpora in opus try to do something like this e.g. [MultiParaCrawl](https://opus.nlpl.eu/datasets/MultiParaCrawl), [MultiCCAligned v1.1](https://opus.nlpl.eu/datasets/MultiCCAligned?pair=de&uk)

Oh, great, thanks for the links, I'll try to use that! I'll be specifically experimenting with ru <-> pt, so I'll keep you posted on how it goes.

> EN->UK model ready !
> 
> Looks pretty solid: https://huggingface.co/quickmt/quickmt-en-uk

Ok, both EN <-> UK models look fantastic, thank you very much! It looks like the uk->en model has the second-most downloads on Huggingface in just 3 days, hope it'll help bring more users and contributors.

> It's tough to beat the "big" Opus models like https://huggingface.co/Helsinki-NLP/opus-mt-tc-big-en-zle. Especially the multi-lingual ones like that one. There definitely seem to be advantages to having multiple related languages mixed in and having backtranslated data like that one does.

Well, I really enjoy your models because they are small, fast (even on CPU), and good enough. Obviously not Google Translate or Deepl quality, but really good. Multilingual ones - I've experienced a lot of issues with them, especially with related languages such as Portuguese, Spanish, and Catalan. I'm also currently running a language detector when translating some texts (a mix of cld, cld3, fasttext and my own naive alphabet-based algorithm). The problem is that these 3 languages are close enough to each other for multilingual LLMs to mix the words and provide bad results (and language detectors literally give unclear results, like 33% portuguese, 33% spanish, 33% catalan, 1% engish, thank you very much for the detection, lol). That's why I prefer exactly 1:1 models and have been searching for them.
